packages/fxa-react/lib/utils.tsx

```diff
@@ -12,3 +14,183 @@ export type FtlMsgProps = {
+ */
+export function checkMessage(
+  bundle: FluentBundle,
+  message: Message,
+  fallbackText: string | null,
+  ftlArgs?: Record<string, FluentVariable>
+) {
+  if (!fallbackText) {
+    throw new Error('Fallback text must be provided.');
+  }
+
+  // nested attributes can happen when we define something like:
+  // `profile-picture =
+  //   .header = Picture`
+  const nestedAttrValues = Object.values(message?.attributes || {});
+
+  if (
+    message === undefined ||
+    (message.value === null && nestedAttrValues.length === 0)
+  ) {
+    throw new Error(`Invalid message. Message has no value.`);
+  }
+
+  if (message.value) {
+    _checkPattern(bundle, message.value, fallbackText, ftlArgs);
+  }
+
+  if (nestedAttrValues) {
+    nestedAttrValues.forEach((nestedAttrValue) =>
+      _checkPattern(bundle, nestedAttrValue, fallbackText, ftlArgs)
+    );
+  }
+}
+
+function _checkPattern(
+  bundle: FluentBundle,
+  pattern: Pattern,
+  fallbackText: string,
+  ftlArgs?: Record<string, FluentVariable>
+) {
+  const errors: Error[] = [];
+  const ftlMsg = bundle.formatPattern(pattern, ftlArgs, errors);
+
+  // Try to render the message, and check for any errors. This will catch things like missing parameters.
+  if (errors.length > 0) {
+    throw new Error(
+      `Errors encountered formatting fluent message. Fluent errors: ${errors
+        .map((x) => x.message)
+        .join('\n')}`
+    );
+  }
+
+  // We allow for .includes because fallback text comes from `textContent` within the `FtlMsg` wrapper which may contain
+  // more than one component and string. Also note that strings must be 'cleaned' because the ftlMsg may have invisible
+  // control characters that can break the includes check.
+  if (!_clean(fallbackText).includes(_clean(ftlMsg))) {
+    throw new Error(
+      `Fallback text does not match Fluent message.\nFallback text: ${fallbackText}\nFluent message: ${ftlMsg}`
+    );
+  }
+
+  let inDomOverlay = false;
+  for (let i = 0; i < ftlMsg.length; i++) {
+    if (!inDomOverlay && ftlMsg[i] === '<') {
+      // Start of dom overlay tag
+      inDomOverlay = true;
+    } else if (inDomOverlay && ftlMsg[i] === '>') {
+      // End of dom overlay tag
+      inDomOverlay = false;
+    } else if (inDomOverlay && ftlMsg[i] === '<') {
+      throw new Error(
+        `Fluent message contains a '<' character inside a dom overlay tag. Check that dom overlay is well formed. Fluent message: ${ftlMsg}`
+      );
+    } else if (!inDomOverlay && ftlMsg[i] === '>') {
+      throw new Error(
+        `Fluent message contains a '>' character that doesn't appear to be part of a dom tag. Either encode with html entity or check dom overlay is well formed. Fluent message: ${ftlMsg}`
+      );
+    } else if (!inDomOverlay && ftlMsg[i] === "'") {
+      throw new Error(
+        `Fluent message contains a straight apostrophe (') and must be updated to its curly equivalent (’). Fluent message: ${ftlMsg}`
```

Comment:
I'm not sure you want these two checks (single and double straight quotes), because some strings have a valid use for them ([example](https://github.com/mozilla/fxa-content-server-l10n/blob/main/locale/en/auth.ftl#L48)).

You'd need to check the string cleaned up from HTML if you don't want to support a list of exclusions.

It would be possible to use [this linter](https://github.com/mozilla-l10n/moz-fluent-linter) for them, and other checks (e.g. hard-coded brand names). If necessary, I can help implement it in the repository. The DOM overlay is out of scope for that, since we have too many implementations :-( 

But I have a more important question: is it correct to assume that all string call will go through this utility (even for localized strings)? Then we definitely need to remove these two checks, because each locale will have their own typography rules.

Reply:
@flodolo, thanks for letting me know about that Linter. I think we could definitely run this on the fluent files we generate. It also seems like this is something that you all should be running on the translated files you provide us.

Also a couple other comments.

1. The ticket explicitly called out these checks on straight quotes, which is why they are in here. Given how/when these checks are actually enabled, I think it makes sense... read on.

1. These checks only kick in if strict mode is enabled. I guess we need to decide when to run with strict mode, but I am of the opinion that we should use strict mode during development and during testing. The idea here is that we would catch any errors early on and during development. For testing / development we'd be operating with an 'en' locale, so I don't think the typography differences between locales are applicable here. Also, if you look at how this code gets invoked, it's always running on the 'en' files and the fallback text. Think of turning on strict mode as enabling developer checks to make sure our fallback text and our extracted en strings are kosher.

2. The example you linked to would actually pass, and not result in error. When parsing, I account for the dom overlay start and end tags. I added this yesterday, so not sure if the extra logic was there at time of comment.

Reply:
Thanks for the explanation.

The reference linter is [already running](https://github.com/mozilla/fxa-content-server-l10n/blob/main/.github/workflows/reference_linter.yaml) in the l10n repo, and there's another one running on the localized files (see top of the [README](https://github.com/mozilla/fxa-content-server-l10n)) ;-)